Docker: entrypoint.sh refactoring

garutilorenzo opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
Build/Testing/Packaging Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Updating entrypoint.sh and Dockerfile to skip some step if tha data directory alredy exist.
In details:
* Modify directory permission
* setup clickhouse user
* process init_files

only if data directory is empty

General entrypoint.sh improvements (use bash funcions)

Detailed description / Documentation draft:
Updating entrypoint.sh and Dockerfile to skip some step if tha data directory alredy exist.
In details:
* Modify directory permission
* setup clickhouse user
* process init_files

only if data directory is empty

General entrypoint.sh improvements (use bash funcions)

garutilorenzo committed:
Updating entrypoint.sh and Dockerfile to skip some step if tha data directory alredy exist. General entrypoint.sh improvements (use bash funcions)

filimonov commented:
1) Motivation is not 100% clear. The only clear improvement i see is ability to source the entrypoint w/o executing it. So you can for example create some adjusted entrypoint which will source the main one. It's nice and common practice. 
2) Backward incompatible? before you could start for example clickhouse-client from the server image. I'm not 100% sure how widely it used by users, but i don think i want to check that this way :) 
3) Separation to functions not accurate, responsibility of some functions are mixed.
4) the PR description unclear: " to skip some step if tha data directory alredy exist" - what did you mean? I don't see any new logic in that part.

garutilorenzo commented:
Hi @filimonov 

> * Motivation is not 100% clear. The only clear improvement i see is ability to source the entrypoint w/o executing it. So you can for example create some adjusted entrypoint which will source the main one. It's nice and common practice.

my intention was to skip some step if the container already exist.
In details if the container starts for the very first time then:

* set correct permission for directory used by clickhouse (setup_dirs)
* create user  (create_clickhouse_user)
* process init_files (docker_process_init_files)
* start clickhouse-server

if the container already exist:

* don't prcess init files (docker_process_init_files):
  * skip statements already executed
  * if clickhouse have already initialized data in this way we prevent data loss (or data overwrite)
* the directory permission was already modified, we skip this step

Actually the user creation set (create_clickhouse_user) can be run always (new container/already existing container)

> * Backward incompatible? before you could start for example clickhouse-client from the server image. I'm not 100% sure how widely it used by users, but i don't think i want to check that this way :)

I thin that clickhouse-client can be run without problem with this revision.
The entrypoint allows to call any command available in the container. The CMD command in Dockerfile will be overwritten by the command passed by the user.
```console
docker run --rm -it garutilorenzo/clickhouse clickhouse-client
ClickHouse client version 20.8.14.4 (official build).
Connecting to localhost:9000 as user default.
Code: 210. DB::NetException: Connection refused (localhost:9000)
```
> * Separation to functions is not accurate, responsibility of some functions are mixed

Actually you are right, i will rearrange funcions in a better way

> * the PR description unclear: " to skip some step if tha data directory alredy exist" - what did you mean? I don't see any new logic in that part.

See response at point 1).
I will give a better description to this PR

garutilorenzo committed:
Merge remote-tracking branch 'upstream/master'

garutilorenzo committed:
Some semplifications

filimonov commented on docker/server/entrypoint.sh:
```diff
@@ -3,164 +3,213 @@
+    DO_CHOWN=1
+    if [ "${CLICKHOUSE_DO_NOT_CHOWN:-0}" = "1" ]; then
+        DO_CHOWN=0
+    fi
```
Move that inside setup_dirs or use CLICKHOUSE_DO_NOT_CHOWN there directly

garutilorenzo replied:
Done

filimonov commented on docker/server/entrypoint.sh:
```diff
@@ -3,164 +3,213 @@
+        USER="$(id -u)"
+        GROUP="$(id -g)"
+        gosu=""
+        DO_CHOWN=0
```
May be just set CLICKHOUSE_DO_NOT_CHOWN

garutilorenzo replied:
Done

filimonov commented on docker/server/entrypoint.sh:
```diff
@@ -3,164 +3,213 @@
-fi
+    declare -g CLICKHOUSE_CONFIG SERVER_ARGS
+    CLICKHOUSE_CONFIG="${CLICKHOUSE_CONFIG:-/etc/clickhouse-server/config.xml}"
+    SERVER_ARGS="$SERVER_ARGS --config-file=$CLICKHOUSE_CONFIG"
```
Config path can (theoretically) have spaces. Pass with quotes

garutilorenzo replied:
I've added double quotes on:

`exec $gosu "$@" "$SERVER_ARGS"`

i think now it's ok.

filimonov commented on docker/server/entrypoint.sh:
```diff
@@ -3,164 +3,213 @@
         <${CLICKHOUSE_USER}>
-          <profile>default</profile>
-          <networks>
+        <profile>default</profile>
```
Why flattening xml?

garutilorenzo replied:
Maybe i'm wrong but for me is equal to the original entrypoint.sh

filimonov replied:
whitespaces at the begining of line

filimonov commented on docker/server/entrypoint.sh:
```diff
@@ -3,164 +3,213 @@
+    if [ "$1" = 'clickhouse-server' ]; then
+        get_variables
+        # Modify dir permission, create users and parse init_files
+        # only if data directory is empyt
```
```suggestion
        # only if data directory is empty
```

filimonov commented on docker/server/entrypoint.sh:
```diff
@@ -3,164 +3,213 @@
+        get_variables
+        # Modify dir permission, create users and parse init_files
+        # only if data directory is empyt
+        if [ -z "$DATABASE_ALREADY_EXISTS" ]; then
```
Ok, this is the change. While user creation is indeed quite stupid for every run, reruning init files and reiniting folders can be desired, i would introduce an option INIT_ON_EVERY_START (or similar) to preserve old behaviour for folders and init files. 

Just in case somebody want old behaviour for some reason.

filimonov commented on docker/server/entrypoint.sh:
```diff
@@ -3,164 +3,213 @@
+    # SERVER_ARGS and gosu variables are defined in get_variables 
+    # and set_user functions.
+    # These variables are defined only if the first argument is clickhouse-server
+    exec $gosu "$@" $SERVER_ARGS
```
Compatibility: Now it will run custom cmd as 'clickhouse' by default. Before it was root.

garutilorenzo replied:
Even in the original image if i run a custom cmd the user was root:

```
docker run --rm -it yandex/clickhouse-server whoami
root
```

garutilorenzo replied:
Same for the new entrypoint.sh

```
docker run --rm -it garutilorenzo/clickhouse whoami
root
```
gosu variable is empty if the command passed to the entrypoint is different from clickhouse-server

filimonov commented:
Please also ensure that https://www.shellcheck.net/ is happy with your changes. 

And it seem we need to start creating test suit for dockerfile / entrypoint itself...

garutilorenzo committed:
Merge remote-tracking branch 'upstream/master'

garutilorenzo committed:
Introduced INIT_ON_EVERY_START env. variable. Use only CLICKHOUSE_DO_NOT_CHOWN insead extra variable DO_CHOWN.

filimonov commented on docker/server/entrypoint.sh:
```diff
@@ -3,164 +3,215 @@
+    # SERVER_ARGS and gosu variables are defined in get_variables.
+    # These variables are defined only if the first argument is clickhouse-server
+    exec $gosu "$@" "$SERVER_ARGS"
+}
```
Hm. But gosu should run the command as clickhouse?

filimonov commented on docker/server/entrypoint.sh:
```diff
@@ -3,164 +3,215 @@
+        # Modify dir permission, create users and parse init_files
+        # only INIT_ON_EVERY_START is equal to 1 (default)
+        # 
+        # To skup this step set INIT_ON_EVERY_START env. variable to 0
```
```suggestion
        # To skip this step set INIT_ON_EVERY_START env. variable to 0
```

filimonov commented on docker/server/entrypoint.sh:
```diff
@@ -3,164 +3,215 @@
+        # only INIT_ON_EVERY_START is equal to 1 (default)
+        # 
+        # To skup this step set INIT_ON_EVERY_START env. variable to 0
+        if [ "$INIT_ON_EVERY_START" = "1" ]; then
```
But now it will not run also on first start?

garutilorenzo replied:
Yes sure, the default of INIT_ON_EVERY_START variable is 1

filimonov commented on docker/server/entrypoint.sh:
```diff
@@ -3,164 +3,215 @@
+        USER="$(id -u)"
+        GROUP="$(id -g)"
+        gosu=""
+        CLICKHOUSE_DO_NOT_CHOWN=0
```
the meaning of flag is opposite

garutilorenzo replied:
Changed CLICKHOUSE_DO_NOT_CHOWN default variable (back to the original value). Now if the container starts with an arbitrary uid this variable is set to 1

filimonov commented on docker/server/entrypoint.sh:
```diff
@@ -3,164 +3,215 @@
+
+get_variables() {
+    declare -g CLICKHOUSE_DO_NOT_CHOWN INIT_ON_EVERY_START
+    CLICKHOUSE_DO_NOT_CHOWN="${CLICKHOUSE_DO_NOT_CHOWN:-1}"
```
default  changed?

filimonov commented on docker/server/entrypoint.sh:
```diff
@@ -3,164 +3,215 @@
+}
+
+get_variables() {
+    declare -g CLICKHOUSE_DO_NOT_CHOWN INIT_ON_EVERY_START
```
are those declare really needed? do they change something? AFAIR in bash vars are global by default.

filimonov commented on docker/server/entrypoint.sh:
```diff
@@ -3,164 +3,215 @@
+	# https://unix.stackexchange.com/a/215279
+	[ "${#FUNCNAME[@]}" -ge 2 ] \
+		&& [ "${FUNCNAME[0]}" = '_is_sourced' ] \
+		&& [ "${FUNCNAME[1]}" = 'source' ]
```
tabs / spaces (we prefer to ident with 4 spaces).

garutilorenzo committed:
Merge remote-tracking branch 'upstream/master'

garutilorenzo committed:
Changed CLICKHOUSE_DO_NOT_CHOWN default to 0, removed declare variables, changed user creation logic, resolved mixed indentation

garutilorenzo commented:
Hi @filimonov 

i've made some changes to the logic of the user creation.
In some circumstances if i start and stop the container without destroying it the user definition file disappeard.
Now i test if the user definition file exist, if not the file will be created.

There is an unexpeted behavior (but i think it was the same with the original entrypoint).
If the container starts with an uid, then the we stop the container and after we start the container with an other uid the entrypoint throws the error "Operation non permitted" when it tries to do the chown command.
I think that is a very remote condition to start a container whit a different uid but it can happen.

Lorenzo

garutilorenzo committed:
Removed debug line

garutilorenzo committed:
Merge remote-tracking branch 'upstream/master'

garutilorenzo committed:
Use sticky bit on clickhouse directories, change owner to clickhouse:clickouse

garutilorenzo committed:
Re-introduced DO_CHOWN variable to keep variables names logical/consistent. Changed gosu logic, now gosu is setup only in _main function.
Simplify logic for chown directories. setup_dirs now is run on evry container starts, this step is skipped only if CLICKHOUSE_DO_NOT_CHOWN is set to 1. Removed SERVER_ARGS to simplify things.

garutilorenzo commented:
Hi @filimonov 

i've made few other changes:

* Dockerfile Use sticky bit on clickhouse directories, change owner to clickhouse:clickouse
* in entrypoint.sh: simplified use of gosu, reintroduced DO_CHOWN variable to to keep variables names logical/consistent, simplified logic for chown directories: setup_dirs now is run on every container starts, this step is skipped only if CLICKHOUSE_DO_NOT_CHOWN is set to 1. Removed SERVER_ARGS to simplify things.

filimonov commented:
> Use sticky bit on clickhouse directories,

Do you see some use cases why it can be useful?

garutilorenzo commented:
> > Use sticky bit on clickhouse directories,
> 
> Do you see some use cases why it can be useful?

I think it can be useful as security feature, but not a mandatory change.
If a container starts with an arbitrary uid, then somone connect to the running container with a different uid it has no permission to delete/change files.

filimonov commented:
Too many changes at a time, hard to review / follow the diff.

Some changes done 'just in case', but may have a negative impact. 

I would prefer to split that PR into 2 (or more) small parts with clear diff & simple motivation each. 

I can do that split on my own - i even almost did it on the top of 46dda06, but now I'm out of sync with your PR again :)

alexey-milovidov commented:
@filimonov
> I can do that split on my own

Yes, it will be very appreciated.
